src/Bicep.Core/Registry/AzureContainerRegistryManager.cs

```diff
@@ -27,104 +35,297 @@ public class AzureContainerRegistryManager
+            if (bicepSources is not null)
             {
-                layer.ResetStream();
-                var layerDescriptor = DescriptorFactory.CreateDescriptor(algorithmIdentifier, layer);
-                layerDescriptors.Add(layerDescriptor);
-
-                layer.ResetStream();
-                _ = await blobClient.UploadBlobAsync(layer.Stream);
+                await PushSourceManifestAsync(blobClient, moduleReference, moduleManifestDescriptor, bicepSources, timestamp);
             }
+        }
 
-            var annotations = new Dictionary<string, string>();
+        private async Task<OciDescriptor> PushModuleManifestAsync(IOciRegistryContentClient blobClient, OciArtifactModuleReference moduleReference, string? artifactType, StreamDescriptor config, string? documentationUri, string? description, StreamDescriptor[] layers, string timestamp)
+        {
+            /* Sample module manifest:
+                {
+                    "schemaVersion": 2,
+                    "artifactType": "application/vnd.ms.bicep.module.artifact",
+                    "config": {
+                      "mediaType": "application/vnd.ms.bicep.module.config.v1+json",
+                      "digest": "sha256:...",
+                      "size": 0
+                    },
+                    "layers": [
+                      {
+                        "mediaType": "application/vnd.ms.bicep.module.layer.v1+json",
+                        "digest": "sha256:...",
+                        "size": 2774
+                      }
+                    ],
+                    "annotations": {
+                      "org.opencontainers.image.description": "module description"
+                      "org.opencontainers.image.documentation": "https://www.contoso.com/moduledocumentation.html"
+                    }
+                  }
+             */
 
+            var annotations = new Dictionary<string, string>();
             if (!string.IsNullOrWhiteSpace(documentationUri))
             {
                 annotations[LanguageConstants.OciOpenContainerImageDocumentationAnnotation] = documentationUri;
             }
-
             if (!string.IsNullOrWhiteSpace(description))
             {
                 annotations[LanguageConstants.OciOpenContainerImageDescriptionAnnotation] = description;
             }
 
-            var manifest = new OciManifest(2, artifactType, configDescriptor, layerDescriptors, annotations);
+            // Adding a timestamp is important to ensure any sources manifests will always point to a unique module
+            //   manifest, even if something in the sources changes that doesn't affect the compiled output (in which
+            //   case, it we don't delete the previous source referrer, you will have two source manifests pointing
+            //   to the same module manifest).
```

Comment:
Not sure if you've already handled this, but we shouldn't leak sources for the same module if it gets republished with or without changes.

I think it should be safe to do that by removing any artifact that refers to the module if its of bicep source type.

Reply:
As discussed, will resolve this by adding timestamps (and comments).  Not in this PR, though.